src/WasmExecutor.cpp

```diff
@@ -0,0 +1,1583 @@
+            internal_assert(!(!prev_used && !r.used));
+            prev_end = start + r.size;
+            prev_used = r.used;
+        }
+        internal_assert(prev_end == total_size) << "prev_end " << prev_end << " total_size "<< total_size<<"\n";
+        bddebug(2)<<"\n";
+#endif
+    }
+
+};
+
+// ---------------------
+
+// Trampolines do not use "_argv" as the suffix because
+// that name may already exist and if so, will return an int
+// instead of taking a pointer at the end of the args list to
+// receive the result value.
+static const char kTrampolineSuffix[] = "_trampoline";
+
+using JITExternMap = std::map<std::string, Halide::JITExtern>;
+
+}  // namespace
+
+#if WITH_V8
+
+#define V8_API_VERSION ((V8_MAJOR_VERSION * 10) + V8_MINOR_VERSION)
+
+static_assert(V8_API_VERSION >= 70,
+              "Halide requires V8 v7.0 or later when compiling WITH_V8.");
+
+namespace Halide {
+namespace Internal {
+namespace {
+
+using namespace v8;
+
+using wasm32_ptr_t = int32_t;
+
+const wasm32_ptr_t kMagicJitUserContextValue = -1;
+
+#if WASM_DEBUG_LEVEL
+void print_object_properties(Isolate *isolate, const Local<Value> &v) {
+    Local<Context> context = isolate->GetCurrentContext();
+    String::Utf8Value objascii(isolate, v);
+    wdebug(0) << *objascii << "\n";
+
+    if (v->IsObject()) {
+        Local<Object> obj = v.As<Object>();
+        Local<Array> properties = obj->GetPropertyNames(context).ToLocalChecked();
+        int len = properties->Length();
+        wdebug(0) << "Number of properties = " << len << ":\n";
+        for(int i = 0 ; i < len ; ++i) {
+            const v8::Handle<v8::Value> key = properties->Get(i);
+            String::Utf8Value str(isolate, key);
+            wdebug(0) << "\t" << i + 1 << ". " << *str << "\n";
+        }
+    }
+}
+#endif
+
+inline constexpr int halide_type_code(halide_type_code_t code, int bits) {
+    return ((int) code) | (bits << 8);
+}
+
+// dynamic_type_dispatch is a utility for functors that want to be able
+// to dynamically dispatch a halide_type_t to type-specialized code.
+// To use it, a functor must be a *templated* class, e.g.
+//
+//     template<typename T> class MyFunctor { int operator()(arg1, arg2...); };
+//
+// dynamic_type_dispatch() is called with a halide_type_t as the first argument,
+// followed by the arguments to the Functor's operator():
+//
+//     auto result = dynamic_type_dispatch<MyFunctor>(some_halide_type, arg1, arg2);
+//
+// Note that this means that the functor must be able to instantiate its
+// operator() for all the Halide scalar types; it also means that all those
+// variants *will* be instantiated (increasing code size), so this approach
+// should only be used when strictly necessary.
+template<template<typename> class Functor, typename... Args>
+auto dynamic_type_dispatch(const halide_type_t &type, Args&&... args) ->
+    decltype(std::declval<Functor<uint8_t>>()(std::forward<Args>(args)...)) {
+
+#define HANDLE_CASE(CODE, BITS, TYPE) \
```

Comment:
I don't think this macro buys you much in terms of clarity. E.g. none of the args are used more than once, and it's a single line.

Reply:
Heh: I think that this is a fine example of one of the few remaining legit uses of the preprocessor; the 'full' line has the differences lost in the noise when I read it, whereas the macro'ed version makes the differences much more obvious.

I'll change this if there's opposition, of course, but IMHO this is clearly more readable than the alternative.

Reply:
I will defer to your judgement.